format/bplist/bplist.go

```diff
@@ -103,83 +103,113 @@ func decodeItem(d *decode.D, p *plist) {
 		n := d.FieldUFn("size", func(d *decode.D) uint64 {
 			return 1 << d.U4()
 		})
-		if n*8 <= 64 {
-			d.FieldU("value", int(n*8))
-		} else {
-			d.FieldUBigInt("value", int(n))
+		switch n {
+		case 1:
+			return d.FieldU8("value")
+		case 2:
+			return d.FieldU16("value")
+		case 4:
+			return d.FieldU32("value")
+		case 8:
+			return d.FieldS64("value")
+		case 16:
+			return d.FieldSBigInt("value", int(n*8))
+		default:
+			d.Errorf("invalid integer size %d", n)
+			return nil
 		}
 	case elementTypeReal:
 		n := 1 << decodeSize(d)
 		d.FieldValueU("size", uint64(n))
-		d.FieldF("value", n*8)
+		return d.FieldF("value", n*8)
 	case elementTypeDate:
 		n := 1 << decodeSize(d, d.AssertU(4, 8))
 		d.FieldValueU("size", uint64(n))
-		d.FieldF("value", n*8, scalar.DescriptionTimeFn(scalar.S.TryActualF, cocoaTimeEpochDate, time.RFC3339))
+		return d.FieldF("value", n*8, scalar.DescriptionTimeFn(scalar.S.TryActualF, cocoaTimeEpochDate, time.RFC3339))
 	case elementTypeData:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldRawLen("value", int64(n*8))
+		return d.FieldRawLen("value", int64(n*8))
 	case elementTypeASCIIString:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUTF8("value", int(n))
+		return d.FieldUTF8("value", int(n))
 	case elementTypeUnicodeString:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUTF16("value", int(n))
+		return d.FieldUTF16("value", int(n))
 	case elementTypeUID:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldUBigInt("value", int(n)).Uint64()
+		return d.FieldUBigInt("value", int(n)).Uint64()
 	case elementTypeArray:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
 			func(d *decode.D) {
-				idx := d.FieldU8("object_index")
-				decodeReference(d, p, idx)
+				idx := d.FieldU("object_index", int(p.t.objRefSize)*8)
+				p.decodeReference(d, idx)
 			})
 	case elementTypeSet:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
 			func(d *decode.D) {
-				idx := d.FieldU8("object_index")
-				decodeReference(d, p, idx)
+				idx := d.FieldU("object_index", int(p.t.objRefSize)*8)
+				p.decodeReference(d, idx)
 			})
 	case elementTypeDict:
 		n := decodeSize(d)
 		d.FieldValueU("size", n)
-		d.FieldStructNArray("entries", "entry", int64(n),
+		return d.FieldStructNArray("entries", "entry", int64(n),
```

Comment:
For the `FieldStructNArray` cases does it make senes to return to *D? is the return value used for dictionary key string? if so an alternative that make be clearer is to have `decodeItem` return `(string, bool)` or  just `string` if empty string is not allowed as key?

Reply:
Good call, going to return a bool indicating whether or not decoded item is a string. Still feels a little awkward but it's better than returning everything and doing type checking. I also verified that empty strings are allowed as dictionary keys.

Reply:
Ah yes it's not even needed to return the string, just decode in the struct? like it. There is some code in fq similar to how it looked in the previous version, would be nice to figure of some decode API for making it nicer Maybe possible with generics but one problem is that with go generics you can't add additional type parameters to a method i think? :(